File: DP/K_stones.cpp

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

vector<int>* answer; 
vector<int>* piles;

int find(int i){
    if ((*answer)[i] != -1) return (*answer)[i];
    int minval = 1;
    for (int val: (*piles)){
        if ((i - val) >= 0) minval = min(minval, find(i-val));
        else break;
    }

        
    int status = 1 - minval;
    (*answer)[i] = status;
    return status;

}
    

int main(){
    int N, K;
    cin >> N >> K;
    vector<int> status(K+1, -1);
    status[0] = 0; //0 means lose if you get it 
    vector<int> pile;
    int val;

    answer = &status;
    piles = &pile;

    for (int i = 0; i < N; i++){
        cin >> val;
        pile.push_back(val);
    }

    sort(pile.begin(), pile.end());

    int win = find(K);
    string thing = (win == 1)? "First": "Second";
    cout << thing;
}

File: DP/K_stones.py
import sys 

lst = sys.stdin.readline().rstrip().split(' ')
N = int(lst[0])
K = int(lst[1])
lst = sys.stdin.readline().rstrip().split(' ')
pile = [int(val) for val in lst]
pile.sort() #sort so then we can decrease a bit better 

sys.setrecursionlimit(10**9)

answer = [-1 for i in range(K + 1)]
answer[0] = 0

def find(i):
    if (answer[i] != -1): 
        return answer[i]
    else:
        minval = 1 
        for val in pile:
            if (i - val) >= 0: #then we can go forth 
                minval = min(minval, find(i-val))
            else:
                break 
    status = 1 - minval 
    answer[i] = status 
    return status 

win = find(K)
print("First" if win else "Second")



